include/deal.II/matrix_free/hanging_nodes_internal.h

```diff
@@ -0,0 +1,997 @@
+
+    void
+    rotate_face(int                                   times,
+                unsigned int                          n_dofs_1d,
+                std::vector<types::global_dof_index> &dofs) const;
+
+    unsigned int
+    line_dof_idx(int          local_line,
+                 unsigned int dof,
+                 unsigned int n_dofs_1d) const;
+
+    void
+    transpose_face(std::vector<types::global_dof_index> &dofs) const;
+
+    void
+    transpose_subface_index(unsigned int &subface) const;
+
+    using cell_iterator        = typename DoFHandler<dim>::cell_iterator;
+    using active_cell_iterator = typename DoFHandler<dim>::active_cell_iterator;
+    const unsigned int n_raw_lines;
+    std::vector<std::vector<std::pair<cell_iterator, unsigned int>>>
+                                     line_to_cells;
+    const std::vector<unsigned int> &lexicographic_mapping;
+    const unsigned int               fe_degree;
+    const DoFHandler<dim> &          dof_handler;
+  };
+
+  namespace internal
+  {
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+    __constant__ double
+      constraint_weights[(CUDAWrappers::mf_max_elem_degree + 1) *
+                         (CUDAWrappers::mf_max_elem_degree + 1)];
+#endif
+
+    // Here is the system for how we store constraint types in a binary mask.
+    // This is not a complete contradiction-free system, i.e., there are
+    // invalid states that we just assume that we never get.
+
+    // If the mask is zero, there are no constraints. Then, there are three
+    // different fields with one bit per dimension. The first field determines
+    // the type, or the position of an element along each direction. The
+    // second field determines if there is a constrained face with that
+    // direction as normal. The last field determines if there is a
+    // constrained edge of a given pair of coordinate planes, but where
+    // neither of the corresponding faces are constrained (only valid in 3D).
+
+    // The element is placed in the 'first position' along *-axis. These also
+    // determine which face is constrained. For example, in 2D, if
+    // constr_face_x and constr_type are set, then x = 0 is constrained.
+    constexpr unsigned int constr_type_x = 1 << 0;
+    constexpr unsigned int constr_type_y = 1 << 1;
+    constexpr unsigned int constr_type_z = 1 << 2;
+
+    // Element has as a constraint at * = 0 or * = fe_degree face
+    constexpr unsigned int constr_face_x = 1 << 3;
+    constexpr unsigned int constr_face_y = 1 << 4;
+    constexpr unsigned int constr_face_z = 1 << 5;
+
+    // Element has as a constraint at * = 0 or * = fe_degree edge
+    constexpr unsigned int constr_edge_xy = 1 << 6;
+    constexpr unsigned int constr_edge_yz = 1 << 7;
+    constexpr unsigned int constr_edge_zx = 1 << 8;
+
+    template <int dim>
+    void
+    setup_constraint_weigths(unsigned int fe_degree)
+    {
+#ifndef DEAL_II_COMPILER_CUDA_AWARE
```

Comment:
This is a bad idea. It's now possible to get two versions of this functions (one from gcc and one from nvcc) and you don't know what's going to happen at link time. When we first wrote the CUDA backend we did the exact same thing and it lead to very subtle bugs.

Reply:
I agree with @Rombur that we should avoid this. If we instantiate this function from both an nvcc compilation unit and a gcc compilation unit and combine them in the library, we don't know which one is chosen. We are good, if this function is recompiled for every compilation unit, though, i.e. we don't provide instantiations in the library.

Reply:
I plan to eliminate this function and move the content to the actual `MatrixFree` classes, since this is FE specific and for multi-component elements, we need multiple of these.

Reply:
Hmmm... I would make sure that the function has a template argument that helps to distinguish between the CUDA and the non-CUDA version to be on the safe side.

Reply:
This is just an intermediate step. Most of the logic is already moved into `ShapeInfo` by #12563. One only, would need to copy the array to the device which can be done simply by the CUDA version of `MatrixFree`. In the CPU version, it is simple to get the interpolation matrix: https://github.com/dealii/dealii/blob/d400159c8a2c176500d718ffc357cb0ca207a627/include/deal.II/matrix_free/evaluation_kernels.h#L4587-L4589

A generalized version of `HangingNodes` can be found in https://github.com/dealii/dealii/blob/d400159c8a2c176500d718ffc357cb0ca207a627/include/deal.II/matrix_free/hanging_nodes_internal.h#L69-L723.

Reply:
> This is just an intermediate step. 

Right but if someone pulls the repo after we merge this PR but before we merge the final version, they may (or not) have a nasty surprise.

Reply:
OK. Now I am guarding the whole function.

Reply:
Hmm... but this constructor is still not fully guarded AFAICT.